### PR TITLE
Factorization and improvement for bounding box computation

### PR DESCRIPTION
**Pull request by Codain**
Hi,

This pull request is a first modest attempt to propose some improvement to OsmAnd, an App I'm using since years now. Please be gentle at reviewing it :).

While crawling the source code on my spare time I noticed that some stuff could probably be implemented a little bit different for maintenance and/or performance reasons. This PR is a first attempt, proposing to group in a single place code used to compute bounding boxes. By doing it I believe that using an helper class will improve clarity and maintenance. Also, it allows some bit of performance improvement, first inclusion being transparent and iteration through the coordinates no longer calling another method. Performance benefit can be arguable, and I didn't measure it, but when thinking about the potential number of points in a route, I'm convinced it is a good direction.

Please note that I'm not fully satisfied about the solution as there is one open topic remaining: There is some code duplication that could be avoided in QuadRectHelper if Location, Node and TargetPoint (and possibly LatLon) would share a common Interface to provide getLatitude/Longitude. But I was not willing to propose such a change in the data structure out of the blue, so if there is an interest for it I propose to add it to this PR, otherwise this PR is already a good step IMHO.

Please share your thoughts.

If things are going well (including welcoming of this PR) I might propose some other improvements from time to time.

Disclaimer: Despite all the care I put in reviewing the PR I was not able to execute and validate the modifications. Consequently, I hope I did not introduce any error.

### CODE REVIEW

File: OsmAnd-java/src/main/java/net/osmand/data/QuadRectHelper.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ */
+public class QuadRectHelper {
+	public static QuadRect createRectForBoundingBoxDetermination() {
+		/* With following initiale values, first position will automatically initialize boundaries of the QuadRect,
+		 * thus avoiding the need to detect first insertion.
+		 */
+		return new QuadRect(Double.MAX_VALUE, -Double.MAX_VALUE, -Double.MAX_VALUE, Double.MAX_VALUE);
+	}
+
+	public static void includeLatLon(@NonNull QuadRect r, double latitude, double longitude) {
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/data/QuadRectHelper.java:23]: This method could be part of QuadRect.

File: OsmAnd-java/src/main/java/net/osmand/data/QuadRectHelper.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	public static void includeLocations(@NonNull QuadRect r, @NonNull List<Location> list) {
+		double latitude;
+		double longitude;
+		
+		for (Location l : list) {
+			latitude = l.getLatitude();
+			longitude = l.getLongitude();
+
+			r.left = Math.min(r.left, longitude);
+			r.right = Math.max(r.right, longitude);
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/data/QuadRectHelper.java:42]: This is copypastet from include latlon

File: OsmAnd-java/src/main/java/net/osmand/data/QuadRectHelper.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	public static void includeNodes(@NonNull QuadRect r, @NonNull List<Node> nodes) {
+		double latitude;
+		double longitude;
+		
+		for (Node n : nodes) {
+			latitude = n.getLatitude();
+			longitude = n.getLongitude();
+
+			r.left = Math.min(r.left, longitude);
+			r.right = Math.max(r.right, longitude);
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/data/QuadRectHelper.java:60]: This is copypaste from includelatlon

File: OsmAnd-java/src/main/java/net/osmand/data/QuadRectHelper.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	 */
+	public static void includeTargetPoints(@NonNull QuadRect r, @NonNull List<TargetPointsHelper.TargetPoint> points) {
+		double latitude;
+		double longitude;
+		
+		for (TargetPointsHelper.TargetPoint p : points) {
+			latitude = p.getLatitude();
+			longitude = p.getLongitude();
+
+			r.left = Math.min(r.left, longitude);
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/data/QuadRectHelper.java:77]: This is copypaste from includelatlon

**Comment by vshcherb** [comment]
Usually a Helper class is always worse than standalone class. I would suggest to extend from QuadRect and make class QuadRectLatLon with methods:
```
boolean insetLocation(double lat, double lon);
boolean isEmpty();
```

All this for-loops will be oneliner and probably not needed to be prepared in advance.

Of course QuadRectLatLon should properly handle < -180 > 180 cause it's important of course going over -180 meridian is an issue cause we need to deal with 2 quadrects, in the end it's tricky. Than it's easier to support with some extent [-360, 360]

**Comment by Codain** [comment]
Hello @vshcherb and thanks for your review, I will prepare a second set of pushes.

Regarding your comments on duplication I fully agree, which is why I proposed in my PR to create an Interface since some are already inheriting from other classes. This way only one method accepting any type of object implementing this interface would have to be maintained. There are also other places where such interface would help. Regarding the name of this interface unfortunately LatLon is taken so I was thinking of `HasLatLon`:

```
public interface HasLatLon {
    public void getLatitude();
    public void getLongitude();
}

```
Can you please give me a feedback on this one? Probably it should be handled in a separate and first PR, unless you prefer that I add it here. Please also share how you see it.

**Comment by vshcherb** [comment]
I and team usually try to follow simple rules:
- No static methods
- No any kind of duplication
- No overengineering (if it's possible to do less code / less classes / less methods) - Better!
- And no need to write code in advance (like in your case non-used methods)
- Of course we apply maximum levels on class / file / function etc.

What I see, quadrect is very generic class which shouldn't has any connection to lat/lon, so it should be LatlonRect which could either inherit or encapsulate

**Pull request closed without merging**